### PR TITLE
Support `AbortController`

### LINKED ISSUE #23: Support `AbortController`
https://developer.mozilla.org/en-US/docs/Web/API/AbortController

### PR DESCRIPTION
**Pull request by jopemachine**
Fixes #23.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -41,6 +41,30 @@ export type Options = {
...
+
+	@example
+	```
+	import pTimeout from 'p-timeout';
+
+	const delayedPromise = () => setTimeout(3000);
+
+	const abortController = new AbortController();
+
+	pTimeout(delayedPromise(), 2000, undefined, {
```
Comment (sindresorhus) [inline index.d.ts:58]: Missing `await`

File: index.d.ts [hunk 2]
```diff
@@ -41,6 +41,30 @@ export type Options = {
 		setTimeout: typeof global.setTimeout;
 		clearTimeout: typeof global.clearTimeout;
 	};
+
+	/**
+	You can abort the promise using [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController).
+
+	**Requires Node.js 16 or later.*
```
Comment (sindresorhus) [inline index.d.ts:48]:
```suggestion
	__Requires Node.js 16 or later.__
```

File: index.js [hunk 3]
```diff
@@ -22,6 +39,18 @@ export default function pTimeout(promise, milliseconds, fallback, options) {
 			...options
 		};
 
+		if (options && options.signal) {
+			const {signal} = options;
+
+			signal.addEventListener('abort', () => {
```
Comment (sindresorhus) [inline index.js:45]: Can you add a TODO to use https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/throwIfAborted when targeting Node.js 18?

File: test.js [hunk 4]
```diff
@@ -88,3 +88,17 @@ test('`.clear()` method', async t => {
 	await promise;
 	t.true(inRange(end(), {start: 0, end: 350}));
 });
+
+test('rejects when calls `AbortController.abort`', async t => {
```
Comment (sindresorhus) [inline test.js:92]:
```suggestion
test('rejects when calling `AbortController#abort()`', async t => {
```

**Comment by sindresorhus** [comment]
Tests are failing

File: index.d.ts (hunk 1 above)
Comment (jopemachine) [reply index.d.ts]: In this example, when using `await pTimeout`, timeout error will be thrown before abortController.abort is executed.
I think this example needs to show how to abort pTimeout promise.

File: index.js (hunk 3 above)
Comment (jopemachine) [reply index.js:45]: I think probably you missed [this comment](https://github.com/sindresorhus/got/pull/2020#discussion_r846141810)

File: test.js [hunk 5]
```diff
@@ -88,3 +88,31 @@ test('`.clear()` method', async t => {
 	await promise;
 	t.true(inRange(end(), {start: 0, end: 350}));
 });
+
+if (globalThis.AbortController !== undefined) {
```
Comment (sindresorhus) [inline test.js:92]: TODO comment for when this can be removed.

File: index.d.ts (hunk 1 above)
Comment (sindresorhus) [reply index.d.ts]: You could either move the settimeout call to be before or use an async-IIFE. Floating promises are bad.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/p-timeout/pull/26#discussion_r874428878

**Pull request merged**